update documentation to expose multi-namespace install modes (#49966)
keycloak/keycloak: docs/guides/operator/installation.adoc

Reviewer: ```suggestion

```

The Operator will watch the namespace where it is installed.

To use a different installation namespace, create an overlay with the desired namespace:

> +

Reviewer: Noticed some extra `+` chars here that render incorrectly.
```suggestion

```

</@profile.ifCommunity>

[[cluster_wide_installation]]
=== Cluster-wide installation

A single {project_name} Operator can reconcile custom resources in all namespaces in the cluster. This avoids running one Operator per namespace while still allowing {project_name} instances to be deployed in separate namespaces.

The Operator supports the standard OLM install modes:

[cols="1,2",options="header"]
|===
| Install mode | Operator watches

| `OwnNamespace`
| The namespace where the Operator is installed

| `SingleNamespace`
| One namespace other than the Operator namespace

| `AllNamespaces`
| All namespaces in the cluster
|===

[WARNING]
====
*Preview support*

Cluster-wide installation is supported for common scenarios, but it is still considered in preview.

* Limitations apply; see <<cluster_wide_limitations>>.
* For non-OLM installations, cluster-wide setup requires additional RBAC configuration.
====

[[cluster_wide_limitations]]
==== Limitations

When the Operator watches all namespaces:

* `spec.podTemplate.spec.serviceAccountName` cannot be used on a `Keycloak` custom resource. The Operator ignores this field and reports a warning.

> * The Operator does not reconcile multiple {project_name} versions in one cluster.

Reviewer: ```suggestion
* The Operator does not support reconciling multiple {project_name} versions in one cluster. A single Operator version supports a single matching {project_name} version.
```

> * You need cluster-admin permissions, or equivalent permissions, to configure cluster-scoped RBAC for non-OLM installations.

Reviewer: This will need to be guarded by the community profile as non-OLM installs are supported upstream only. In other words, this bullet point must be excluded downstream.

Author: done

Reviewer: The bullet point should've stayed, just be wrapped by `profile.ifCommunity`.

Reviewer: The bullet-point is still missing. However, thinking about it more, you need cluster-admin permissions even for OLM installations. So the bullet point should be:
```
* You need cluster-admin permissions, or equivalent permissions.
```

==== Installing with OLM

> To install in `AllNamespaces` mode, create an OperatorGroup without `spec.targetNamespaces`, subscribe to the Operator, and approve the install plan.

Reviewer: This makes an impression that the cluster-wide OLM install is possible only via CLI. We should mention the web console way as well.

Author: how do you want me to put that?

Reviewer: We should split the instructions for CLI and web console, similarly how we have it in the chapters above for installing using OLM.

  | sed '/resourceVersion:/d;/uid:/d;/creationTimestamp:/d;/generation:/d' \
  | sed 's/kind: Role/kind: ClusterRole/' \
  | sed '/^  namespace:/d' \
  | kubectl apply -f -

kubectl delete role keycloak-operator-role -n $OPERATOR_NAMESPACE

kubectl get rolebinding keycloak-operator-role-binding -n $OPERATOR_NAMESPACE -o yaml \
  | sed '/resourceVersion:/d;/uid:/d;/creationTimestamp:/d;/generation:/d' \
  | sed 's/kind: RoleBinding/kind: ClusterRoleBinding/' \
  | sed 's/kind: Role/kind: ClusterRole/' \
  | sed '/^  namespace:/d' \
  | kubectl apply -f -

kubectl delete rolebinding keycloak-operator-role-binding -n $OPERATOR_NAMESPACE
----

. Configure the Operator to watch all namespaces and restart it.

+
[source,bash,subs="attributes+"]
----
kubectl set env deployment/keycloak-operator \
  -n $OPERATOR_NAMESPACE \
  QUARKUS_OPERATOR_SDK_NAMESPACES=JOSDK_ALL_NAMESPACES

kubectl rollout restart deployment/keycloak-operator -n $OPERATOR_NAMESPACE
----

. Create `Keycloak` custom resources in any namespace. The Operator reconciles resources cluster-wide.
</@profile.ifCommunity>

=== Installing Multiple Operators

> In most cases, a single Operator installed in cluster-wide mode is simpler than installing multiple Operators. Install multiple Operators only when you need separate Operator instances, for example to run different Operator versions in parallel.

Reviewer: ```suggestion
In most cases, a single Operator installed in cluster-wide mode is simpler than installing multiple Operators. Install multiple Operators only when you need separate Operator instances, for example to run different {project_name} or Operator versions in parallel.
```

# if you have existing keycloaks, restart the operator after patching the clusterrolebinding
kubectl rollout restart -n custom-namespace Deployment/keycloak-operator
----

For cluster-wide installation without OLM, see <<installing_cluster_wide_with_kubectl>>.
</@profile.ifCommunity>

[[multi_namespace_installation]]
=== Multi-namespace installation

A single {project_name} Operator can reconcile custom resources in more than one namespace. This avoids running one Operator per namespace while still allowing {project_name} instances to be deployed in separate namespaces.

The Operator supports the standard OLM install modes:

[cols="1,2",options="header"]
|===
| Install mode | Operator watches

| `OwnNamespace`
| The namespace where the Operator is installed

| `SingleNamespace`
| One namespace other than the Operator namespace

| `MultiNamespace`
| Multiple namespaces listed on the OperatorGroup

| `AllNamespaces`
| All namespaces in the cluster
|===

[WARNING]
====
*Experimental support*

> Multi-namespace and cluster-wide installation is supported for common scenarios, but it is still considered experimental.

Reviewer 2: @vmuzikar do you want this to say preview as well for consistency with the non-olm pr?

Reviewer: Yes, please :)

Author: done

* Limitations apply; see <<multi_namespace_limitations>>.
* For non-OLM installations, cluster-wide setup requires additional RBAC configuration.
====

[[multi_namespace_limitations]]
==== Limitations

When the Operator watches more than one namespace:

* `spec.podTemplate.spec.serviceAccountName` cannot be used on a `Keycloak` custom resource. The Operator ignores this field and reports a warning.
* The Operator does not filter namespaces by labels, nor does it reconcile multiple {project_name} versions in one cluster.
* You need cluster-admin permissions, or equivalent permissions, to configure cluster-scoped RBAC for non-OLM installations.

==== Installing with OLM

> To install in `MultiNamespace` or `AllNamespaces` mode, create an OperatorGroup with the desired namespace selection, subscribe to the Operator, and approve the install plan.

Reviewer 2: I would leave MultiNamespace instructions out of the guide - that install mode is effectively going away.

I also don't see any forward support for namespace selection by label, nor is that currently supported in the java operator sdk.

So it seems better to not even mention the possibility of doing this.

Should we even expose the MultiNamespace install mode in the CSV metadata? cc @vmuzikar

Reviewer: > Should we even expose the MultiNamespace install mode in the CSV metadata?

As discussed on the meeting yesterday, I lean towards no.

===== Watching multiple selected namespaces

Create an OperatorGroup that lists every namespace the Operator should watch:

[source,yaml]
----
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: keycloak-operator-group
  namespace: <operator-namespace>
spec:
  targetNamespaces:
    - tenant-a
    - tenant-b
----

Approve the install plan if you are using manual approval, then verify that the Operator pod is running.

===== Watching all namespaces (cluster-wide)

Create a global OperatorGroup in the namespace where the Operator is installed by omitting `spec.targetNamespaces`:

[source,yaml]
----
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: keycloak-operator-group
  namespace: <operator-namespace>
spec: {}
----

> Approve the install plan if you are using manual approval, then verify that the Operator pod is running.

Reviewer 2: We recommend only using manual approval.

> If the Operator is already installed in single-namespace mode, you can switch to cluster-wide mode by removing `spec.targetNamespaces` from the OperatorGroup:

Reviewer 2: This should be removed. It's not recommended to simply switch from single namespace to cluster-wide -  a cluster-wide operator should be in it's own namespace.

[source,bash]
----
kubectl patch operatorgroup <operator-group-name> -n <operator-namespace> --type json \
  --patch '[{"op":"remove","path":"/spec/targetNamespaces"}]'
kubectl rollout restart deployment/keycloak-operator -n <operator-namespace>
----

<@profile.ifCommunity>

> [[installing_cluster_wide_with_kubectl]]

Reviewer 2: This is fine for a placeholder, but we'll mostly remove / rewrite this based upon the other PR.

[WARNING]
====
This installation path is in preview and applies only to non-OLM installations.

The published `kubernetes.yml` manifest installs namespace-scoped RBAC and configures the Operator to watch its own namespace only. Cluster-wide installation requires cluster-scoped RBAC and an environment variable that enables watching all namespaces.
====

. Install the CRDs and Operator as described in <<installing_with_kubectl>>.

. Promote the Operator role and role binding to cluster-scoped resources.

+
[source,bash,subs="attributes+"]
----
OPERATOR_NAMESPACE=keycloak

kubectl get role keycloak-operator-role -n $OPERATOR_NAMESPACE -o yaml \
  | sed '/resourceVersion:/d;/uid:/d;/creationTimestamp:/d;/generation:/d' \
  | sed 's/kind: Role/kind: ClusterRole/' \
  | sed '/^  namespace:/d' \
  | kubectl apply -f -

kubectl delete role keycloak-operator-role -n $OPERATOR_NAMESPACE

kubectl get rolebinding keycloak-operator-role-binding -n $OPERATOR_NAMESPACE -o yaml \
  | sed '/resourceVersion:/d;/uid:/d;/creationTimestamp:/d;/generation:/d' \
  | sed 's/kind: RoleBinding/kind: ClusterRoleBinding/' \
  | sed 's/kind: Role/kind: ClusterRole/' \
  | sed '/^  namespace:/d' \

>   | kubectl apply -f -

Reviewer 2: Can be ignored, these instructions will be refined later.

>   | kubectl apply -f -

Reviewer 3: > Can be ignored, these instructions will be refined later.

Are you sure it will be refined before 26.7 is released? Maybe it is tracked somewhere and I just don't know about it? I have limited understanding on this topic, but this Copilot comment seems valid. It sounds like if we don't refine it soon, users following this instruction will get an error.

. Search for "keycloak" and select the {project_name} Operator from the *fast* channel.

. On the install screen, select *All namespaces on the cluster* as the installation mode.

. Select *Manual* approval for updates.

. Complete the installation and approve the install plan when prompted.

. Verify that the Operator pod is running in the operator namespace.

===== Using the CLI

Create a global OperatorGroup in the namespace where the Operator is installed by omitting `spec.targetNamespaces`:

[source,yaml]
----
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: keycloak-operator-group
  namespace: <operator-namespace>
spec: {}
----

Create a Subscription with `installPlanApproval: Manual`:

[source,yaml]
----
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: keycloak-operator
  namespace: <operator-namespace>
spec:
  channel: fast
  name: keycloak-operator
  source: <catalog-source>
  sourceNamespace: <catalog-namespace>
  installPlanApproval: Manual
----

Approve the install plan, then verify that the Operator pod is running.

<@profile.ifCommunity>
[[installing_cluster_wide_with_kubectl]]

> ==== Installing cluster-wide with kubectl

Reviewer: @edewit Thank you for the update. There's one last thing from my perspective.

https://github.com/keycloak/keycloak/pull/46980 is ready and will be merged sooner than this. Can you please update this chapter for installing cluster-wide with kubectl to match the current state of https://github.com/keycloak/keycloak/pull/46980?

Additionally, we'll need to mention kustomize in the release notes – both in the context of all-namespaces mode, as well as in the context of generally new deployment method (that the Operator can now be deployed using kustomize).

Reviewer: Actually, https://github.com/keycloak/keycloak/pull/46980 already include kustomize instructions for cluster-wide. So let's just wait until it's merged, rebase this PR and update the docs so cluster-wide instructions for OLM and kustomize are in the same chapter.

Author: done

----

For cluster-wide installation without OLM, see <<installing_cluster_wide_with_kubectl>>.
</@profile.ifCommunity>

[[cluster_wide_installation]]
=== Cluster-wide installation

A single {project_name} Operator can reconcile custom resources in all namespaces in the cluster. This avoids running one Operator per namespace while still allowing {project_name} instances to be deployed in separate namespaces.

The Operator supports the following OLM install modes (`MultiNamespace` is not supported):

[cols="1,2",options="header"]
|===
| Install mode | Operator watches

| `OwnNamespace`
| The namespace where the Operator is installed

| `SingleNamespace`
| One namespace other than the Operator namespace

| `AllNamespaces`
| All namespaces in the cluster
|===

Install the cluster-wide Operator in a dedicated namespace rather than converting an existing single-namespace installation.

[WARNING]
====
*Preview support*

Cluster-wide installation is supported for common scenarios, but it is still considered in preview.

* Limitations apply; see <<cluster_wide_limitations>>.
<@profile.ifCommunity>
* For non-OLM installations, cluster-wide setup requires additional RBAC configuration.

> </@profile.ifCommunity>

Reviewer: Sorry, I noticed this just now, it was like it even in the previous iterations of this PR.

This box is containing info that cluster-wide is preview. Yet, it includes additional two bullet points that are irrelevant to the preview state. I'd remove both of the bullet points and keep just:
```
Cluster-wide installation is supported for common scenarios, but it is still considered in preview.
```

A single {project_name} Operator can reconcile custom resources in all namespaces in the cluster. This avoids running one Operator per namespace while still allowing {project_name} instances to be deployed in separate namespaces.

> * You need cluster-admin permissions, or equivalent permissions.

Reviewer: My [suggestion](https://github.com/keycloak/keycloak/pull/49966#discussion_r3490603954) was to add this bullet point back to the Limitations chapter (as it was there before and then was mistakenly removed). Here it seems a bit random.

Author: my mistake moved it to line 177
